### PR TITLE
Add documentation page on GitLab CI

### PR DESCRIPTION
**Pull request by CLOVIS-AI**
### Context

**This page is for the Gradle community site.**

GitLab is a popular CI provider, much like GitHub Actions, Jenkins, etc.

The addition of this page was suggested in the [Gradle Slack](https://gradle-community.slack.com/archives/CAHSN3LDN/p1709368339364999).

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html). **→ I wrote all the code.**
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective. **→ There is no code changes, thus no test changes.**
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic. **→ There is no code changes, thus no test changes.**
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`. **→ There is no code changes.**

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+== Setup a Gradle project on GitLab
+
+If you have an existing Gradle project hosted on GitLab, then you can skip this step and move directly to <<#sec:configure_gitlab,Configure GitLab CI>>.
+
+If not, follow these step to initialize a new Gradle project on GitLab.
+
+=== Create a new GitLab repository for your project
+
+Via the GitLab user interface, create a new repository named `gitlab-ci-gradle-sample`.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:46]:
```suggestion
Create a new repository named `gitlab-ci-gradle-sample` via the GitLab user interface.
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+$ gradle init
+$ git add .
+$ git commit -m "Initial commit"
+$ git push
+----
+
+=== Enable Build Scan™ publishing
+
+https://scans.gradle.com[Gradle Build Scans] are a great way to view your build results, and provide valuable insights into your build.
+In order to publish Build Scans from GitHub Actions, you'll need to pre-approve the Terms & Conditions.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:77]:
```suggestion
To publish Build Scans from GitHub Actions, you'll need to pre-approve the Terms & Conditions.
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+To do so, add the following content to the top of your `settings.gradle[.kts]` file. The "CI" environment variable is set by GitLab CI:
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:79]:
```suggestion
To do so, add the following content to the top of your `settings.gradle[.kts]` file. GitLab CI sets the "CI" environment variable:
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            termsOfServiceAgree = "yes"
+        }
+    }
+}
+```
+
+=== Test building the project
+
+The project uses the <<gradle_wrapper.adoc#gradle_wrapper_reference,Gradle Wrapper>> for building the project.
+It is a recommended practice for any Gradle project as it enables your project to built on CI without having to install the Gradle runtime.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:99]:
```suggestion
It is a recommended practice for any Gradle project as it enables your project to build on CI without installing the Gradle runtime.
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+  artifacts:
+    reports:
+      junit: '**/build/test-results/**/TEST-*.xml'
+----
+
+This file defines a single job that executes the `./gradlew build` command.
+
+* The `image` field defines which Docker image the job will run in. Thanks to the Gradle Wrapper, only a JDK is necessary.
+* The `artifacts` field defines files generated by the job that GitLab CI should save. In particular, saving the JUnit report generated by the Gradle test tasks allows displaying test results in the Merge Request or Pipeline pages directly.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:150]:
```suggestion
* The `artifacts` field defines files generated by the job that GitLab CI should save. In particular, saving the JUnit report generated by the Gradle test tasks allows test results to be displayed directly in the Merge Request or Pipeline pages.
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+Commit the changes and push to the repository:
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:152]:
```suggestion
Commit the changes and push them to the repository:
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+[listing.terminal.sample-command]
+----
+$ git add .
+$ git commit -m "Add GitLab CI pipeline"
+$ git push
+----
+
+== View the GitLab CI results
+
+Once this pipeline file is pushed, you should immediately see the pipeline execution in the GitLab CI page for your repository (eg https://gitlab.com/username/repository/-/pipelines).
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:163]:
```suggestion
Once this pipeline file is pushed, you should immediately see the pipeline execution in the GitLab CI page for your repository (e.g., https://gitlab.com/username/repository/-/pipelines).
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+Any subsequent push to the repository will trigger the pipeline to run.
+
+== Enable caching
+
+By default, no files are persisted between CI runs, so each job must download all dependencies again. We can configure GitLab to cache some files between jobs. This is especially efficient when running your own GitLab CI runner, as the cache is stored locally on the runner.
+
+CAUTION: The publicly available shared runners store their cache in a remote location. As the cache grows, downloading the cache archive, extracting it, and uploading it again at the end of the job may quickly take more time than not using any cache at all. For this reason, we recommend only caching the Gradle Wrapper. Prefer configuring a link:https://docs.gradle.org/current/userguide/build_cache.html[Gradle Remote Build Cache] for caching everything else.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:170]:
```suggestion
CAUTION: The publicly available shared runners store their cache in a remote location. As the cache grows, downloading the cache archive, extracting it, and uploading it again at the end of the job may take more time than not using any cache. For this reason, we recommend only caching the Gradle Wrapper. For caching everything else, see configuring a link:https://docs.gradle.org/current/userguide/build_cache.html[Gradle Remote Build Cache].
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+Learn more about building Gradle projects with GitLab CI:
+
+* https://docs.gitlab.com/ee/ci/[GitLab CI documentation]
+* https://docs.gitlab.com/ee/ci/yaml/[.gitlab-ci.yml syntax reference]
+* https://docs.gitlab.com/ee/ci/variables/predefined_variables.html[Predefined variables]
+* https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html[Avoiding duplication between multiple jobs]
+
+== Summary
+
+Executing Gradle builds on CI can be set up and configured with just a handful of steps.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:209]:
```suggestion
Setting up and configuring Gradle builds on CI is straightforward, requiring just a few steps. 
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+The benefit of receiving fast feedback clearly speaks for itself.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:210]:
```suggestion
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+GitLab CI provides a simple, convenient mechanism to setup CI for any Gradle project hosted on GitLab.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:211]:
```suggestion
GitLab CI offers a simple and convenient mechanism to set up CI for any Gradle project hosted on GitLab.
```

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[build_gitlab]]
+= Executing Gradle builds on GitLab CI
+
+TIP: Top engineering teams using GitHub Actions have been able to reduce CI build time by up to 90% by using the Gradle Build Cache. https://gradle.org/training/#build-cache-deep-dive[Register here] for our Build Cache training session to learn how your team can achieve similar results.
```
[the remark this answers is not in the document]
Comment (CLOVIS-AI) [reply platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:18]: This is a copy-paste of the similar header that is present in all build tooling pages. If it should be changed, I do not believe it should happen within this PR.

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc (hunk 10 above)
Comment (CLOVIS-AI) [reply platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:210]: This sentence is present in all other pages.

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+https://scans.gradle.com[Gradle Build Scans] are a great way to view your build results and provide valuable insights into your build.
+To publish Build Scans from GitLab CI, you'll need to pre-approve the Terms & Conditions.
+
+To do so, add the following content to the top of your `settings.gradle[.kts]` file. GitLab CI sets the "CI" environment variable:
+```
+plugins {
+    id("com.gradle.enterprise") version("3.16.2")
+}
+
+gradleEnterprise {
```
Comment (oleg-nenashev) [inline platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:85]: From what I know, it is expected to be the named to develocity soon in the plugin

**Review by oleg-nenashev — COMMENTED** [review]
It is definitely a great start. I would probably have a list of top level advanced recommendations similar to what you shared on Slack. It is very important to have a starter guide as you created, and it's much appreciated. For advanced users, I think a brief list of tips like you did would be a great addition even if you don't spend too much time documenting them

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc (hunk 13 above)
Comment (CLOVIS-AI) [reply platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:85]: Can I assume that whenever it happens, someone will make the change in all page docs, and thus this one as well? I'd prefer keeping this sample the same as the one in the GitHub page for simplicity's sake.

**Comment by CLOVIS-AI** [comment]
> For advanced users, I think a brief list of tips like you did would be a great addition

If you are happy with the PR in its current state, would it be ok with you to merge this, and then I can create a follow-through PR to discuss the tips? I would prefer keeping PRs small if it's alright with you.

**Comment by oleg-nenashev** [comment]
> If you are happy with the PR in its current state, would it be ok with you to merge this, and then I can create a follow-through PR to discuss the tips? I would prefer keeping PRs small if it's alright with you.

I am +1

File: platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc (hunk 13 above)
Comment (oleg-nenashev) [reply platforms/documentation/docs/src/docs/userguide/reference/ci-systems/gitlab.adoc:85]: Yes. If not, this entry will not be the biggest problem for sure

**Review by oleg-nenashev — APPROVED** [review]
I am +1 for merging. so that we have a foundation to work with

**Review by lkasso — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Sorry for the delay. I am still waiting for the community sub-site launch approval. Once I have it, I will move this PR and all other CI documents there

**Comment by oleg-nenashev** [comment]
Just an update: we published the first version of the Gradle Cookbook PoC, and we have already moved the CI integrations pages there. @CLOVIS-AI I will move this page as an experiment if you do not mind

**Comment by CLOVIS-AI** [comment]
> Just an update: we published the first version of the Gradle Cookbook PoC, and we have already moved the CI integrations pages there. @CLOVIS-AI I will move this page as an experiment if you do not mind

No problem, go ahead. Is this available publicly? It's the first time I hear of it.

**Comment by oleg-nenashev** [comment]
@CLOVIS-AI We're just starting put it is open source and public, e.g. https://community.gradle.org/cookbook/ci/jenkins/ . See the scope and the references are on https://github.com/gradle/community/issues/42

**Comment by oleg-nenashev** [comment]
Hi @CLOVIS-AI . The pull request is created: https://github.com/gradle/community/pull/45
I will go ahead and close this pull request is that's fine with you

**Comment by CLOVIS-AI** [comment]
> Hi @CLOVIS-AI . The pull request is created: [gradle/community#45](https://github.com/gradle/community/pull/45) I will go ahead and close this pull request is that's fine with you

Go ahead :)

Out of curiosity; what is the plan for the main documentation website? Will it be entirely replaced by this new one? If not, what will be in the official site, and what will be in the community one? It seems strange to split information in two different places, especially with how different they look.

**Comment by oleg-nenashev** [comment]
OK, the page is now stable. I will keep tinkering with layout improvements, etc.

**Pull request closed without merging**